cmd/dockerd/daemon.go

```diff
@@ -391,18 +395,24 @@ func setOTLPProtoDefault() {
 	}
 }
 
-func newTracerProvider(ctx context.Context) *sdktrace.TracerProvider {
+func newTracerProvider(ctx context.Context) (trace.TracerProvider, func(context.Context) error) {
 	opts := []sdktrace.TracerProviderOption{
 		sdktrace.WithResource(resource.Default()),
 		sdktrace.WithSyncer(detect.Recorder),
 	}
 
-	if exp, err := detect.NewSpanExporter(ctx); err != nil {
+	exp, err := detect.NewSpanExporter(ctx)
+	if err != nil {
 		log.G(ctx).WithError(err).Warn("Failed to initialize tracing, skipping")
-	} else if !detect.IsNoneSpanExporter(exp) {
-		opts = append(opts, sdktrace.WithBatcher(exp))
+		return noop.NewTracerProvider(), nil
```

Comment:
Curious, as the basic issue is quite similar; would it make sense to apply the workaround in the same location as we did for that previous OTEL memory leak, or is it not possible to set them up in the same location? (sorry didn't look in-depth, mostly considering that it would be nice to have such workarounds close together);

- https://github.com/moby/moby/pull/48690

Reply:
This one is not really a workaround for the OTEL itself. It fixes our code which created and set a full-fledged tracer provider with no export target, rather than just NOT calling `SetTracerProvider` at all, or explicitly setting it to a noop provider.